File: server/src/routes/node.ts
import { NodeController } from '@/controllers';
import { NodeRoute } from '@/enums';
import { NodeMiddlewares } from '@/middlewares';
import Elysia from 'elysia';

export const NodeRouter = new Elysia({ name: 'node.router' })
    .use(NodeController)
    .use(NodeMiddlewares)
    .onError(({ error }) => new Response(error.message))
    .get(NodeRoute.RETRIEVE, ({ retrieveNodes }) => retrieveNodes())
    .get(NodeRoute.FIND_BY_ID, ({ findNodeById }) => findNodeById(), {
        beforeHandle: ({ validateParams }) => validateParams(),
    })
    .post(NodeRoute.CREATE, ({ createNode }) => createNode(), {
        beforeHandle: ({ validatePayload }) => validatePayload(),
    })
    .put(NodeRoute.UPDATE, ({ updateNode }) => updateNode(), {
        beforeHandle: [
            ({ validateParams }) => validateParams(),
            ({ validatePayload }) => validatePayload(),
        ],
    })
    .delete(NodeRoute.DELETE, ({ deleteNode }) => deleteNode(), {
        beforeHandle: ({ validateParams }) => validateParams(),
    });

File: server/internal/routers/node.go
package routers

import (
	"sukejuuru/internal/controllers"
	"sukejuuru/internal/middlewares"

	"github.com/go-chi/chi/v5"
)

func UseNodeRouter(r chi.Router) {
	r.Get("/nodes", controllers.UseNode().RetrieveNodes)
	r.With(middlewares.UseNode().ValidateParams).Get("/nodes/{id}", controllers.UseNode().RetrieveNodeById)
	r.With(middlewares.UseNode().ValidatePayload).Post("/nodes", controllers.UseNode().AddNode)
}
